Collection view and serializer for upload endpoint

nbilenko opened the pull request:
---
name: Collection view and serializer for upload endpoint
about: Upload
title: Collection view and serializer for upload endpoint
labels: MVP, backend, upload-view

---

## Checklist
- [x] Add description
- [x] Reference open issue pull request addresses
- [x] Pass functional tests
  - complete on the local machine with `make local.shell` `make test`
- [x] Request code review
  - Please allow **36 hours** from opening a pull request before merging a pull request- even if it has already recieved an approving review.
- [x] Address comments on code and resolve requested changes
- [x] Merge own code

## Description
Issue: https://github.com/openoakland/woeip/issues/145

Adds create() method to the Collection view and serializer that processes a POST request to collection. Handles metadata object (containing `starts_at` and `ends_at` fields) and file data (passed as files in the request), initializes new Collection and associated CollectionFiles.

nbilenko committed:
adding file processing to collection serializer

nbilenko committed:
fix typo

nbilenko committed:
removing sensor from collection creation in serializer

nbilenko committed:
rebase and missing migration

nbilenko committed:
adds HTTP responses to collection view

r-b-g-b reviewed (changes requested):
Probably worth including a test or two to this, but after that I think it's good to go!

ifed3 commented on api/woeip/apps/air_quality/serializers.py:
```diff
@@ -56,6 +56,19 @@ class Meta:
+                collection=collection,
+                file=file,
+            )
+        return collection
```
@nbilenko, what do you think about using a map function to write a bit cleaner code?

```python
class Meta:
    model = Collection
    fields = ["starts_at", "ends_at"]

def create(self, validated_data):
    collection = Collection.objects.create(**validated_data)
    files = self.context.get('view').request.FILES
    map(lambda file: self.__create_files(collection, file), files)
    return collection

def __create_files(self, collection, file):
    CollectionFile.objects.create(
        collection=collection,
        file=file
    )
```

ifed3 commented on api/woeip/apps/air_quality/views.py:
```diff
@@ -38,6 +39,13 @@ def data(self, request, pk=None):
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```
@nbilenko, there is a cool serializer.is_valid use case with automatic validation http error response. So we can use that as a guard statement here

```python
def create(self, request, *args, **kwargs):
    serializer = self.get_serializer(data=request.data)
    serializer.is_valid(raise_exception=True)
    serializer.save()
    return Response(serializer.data, status=status.HTTP_201_CREATED)
```

ifed3 reviewed (approved):
@nbilenko the code is really clean 👍 and I want to move it forward! I added some suggestions on make the code more readable, so feel free to incorporate.

@r-b-g-b, at this stage, we can focus less on unit testing and do more with integration testing and documentation as testing. We can use [api documetation](https://www.django-rest-framework.org/topics/documenting-your-api/) to explore any endpoint route to test if the responses and errors are as expected

nbilenko committed:
Modified collection creation serializer and added tests

nbilenko committed:
update serializer validation

nbilenko commented:
Thanks for the reviews!

@r-b-g-b Thanks for asking for tests! I should have included them in the first place - writing them ended up uncovering a few different issues, so I'm glad you kept me honest and lets indeed continue to keep coverage thorough.

In addition to tests, I added temporary media file storage logic to the test set up - otherwise django creates a bunch of files in the media dir during testing that don't get removed.

@ifed3 that's a cool validation feature, updated. Please note that the upload endpoint expects the following from the frontend (as data json):
```
{"files": [{"file_name": "gps_file.log", "file_data": b"gpsfiledata"}, {"file_name": "dustrak_file.csv", "file_data": b"dustrakfiledata"}],
 "starts_at": <datetime_obj>,
 "ends_at": <datetime_obj>,
}
```

r-b-g-b reviewed (approved):
Looks good to me!
